watertap/unit_models/tests/test_electroNP_ZO.py

```diff
@@ -0,0 +1,263 @@
+        m.fs.unit.inlet.conc_mass_comp[0, "X_AUT"].fix(EPS * units.mg / units.liter)
+        m.fs.unit.inlet.conc_mass_comp[0, "X_MeOH"].fix(EPS * units.mg / units.liter)
+        m.fs.unit.inlet.conc_mass_comp[0, "X_MeP"].fix(EPS * units.mg / units.liter)
+        m.fs.unit.inlet.conc_mass_comp[0, "X_TSS"].fix(EPS * units.mg / units.liter)
+        m.fs.unit.inlet.conc_mass_comp[0, "S_K"].fix(EPS * units.mg / units.liter)
+        m.fs.unit.inlet.conc_mass_comp[0, "S_Mg"].fix(EPS * units.mg / units.liter)
+
+        # Alkalinity was givien in mg/L based on C
+        m.fs.unit.inlet.alkalinity[0].fix(61 / 12 * units.mmol / units.liter)
+
+        # Unit option
+        m.fs.unit.energy_electric_flow_mass.fix(0.044 * units.kWh / units.kg)
+        m.fs.unit.magnesium_chloride_dosage.fix(0.388)
+
+        return m
+
+    @pytest.mark.unit
+    def test_dof(self, ElectroNP_frame):
+        m = ElectroNP_frame
+        assert degrees_of_freedom(m) == 0
+
+    @pytest.mark.unit
+    def test_units(self, ElectroNP_frame):
+        assert_units_consistent(ElectroNP_frame)
+
+    @pytest.mark.unit
+    def test_calculate_scaling(self, ElectroNP_frame):
+        m = ElectroNP_frame
+        calculate_scaling_factors(m)
+
+    @pytest.mark.solver
+    @pytest.mark.skipif(solver is None, reason="Solver not available")
+    @pytest.mark.component
+    def test_initialize(self, ElectroNP_frame):
+        initialization_tester(ElectroNP_frame)
+
+    @pytest.mark.solver
+    @pytest.mark.skipif(solver is None, reason="Solver not available")
+    @pytest.mark.component
+    def test_solve(self, ElectroNP_frame):
+        m = ElectroNP_frame
+        results = solver.solve(m)
+
+        # Check for optimal solution
+        assert_optimal_termination(results)
+
+    @pytest.mark.solver
+    @pytest.mark.skipif(solver is None, reason="Solver not available")
+    @pytest.mark.component
+    def test_conservation(self, ElectroNP_frame):
+        m = ElectroNP_frame
+        assert (
+            abs(
+                value(
+                    m.fs.unit.inlet.flow_vol[0] * m.fs.properties.dens_mass
+                    - m.fs.unit.treated.flow_vol[0] * m.fs.properties.dens_mass
+                    - m.fs.unit.byproduct.flow_vol[0] * m.fs.properties.dens_mass
+                )
+            )
+            <= 1e-6
+        )
+        for j in m.fs.properties.solute_set:
+            assert 1e-6 >= abs(
+                value(
+                    m.fs.unit.inlet.flow_vol[0] * m.fs.unit.inlet.conc_mass_comp[0, j]
+                    - m.fs.unit.treated.flow_vol[0]
+                    * m.fs.unit.treated.conc_mass_comp[0, j]
+                    - m.fs.unit.byproduct.flow_vol[0]
+                    * m.fs.unit.byproduct.conc_mass_comp[0, j]
+                )
+            )
+
+    @pytest.mark.solver
+    @pytest.mark.skipif(solver is None, reason="Solver not available")
+    @pytest.mark.component
+    def test_solution(self, ElectroNP_frame):
```

Comment:
Is it worth listing out the solution for the byproduct stream too?

Reply:
Since we don't have any byproduct, so I don't think we need to list them all.

Reply:
I thought the byproduct stream would include the S_NH4 and S_PO4 that got removed?

Reply:
I agree with @MarcusHolly. The treated stream would be fed back to the headworks of the plant or right before the activated sludge process, while the byproduct stream would comprise the recovered nutrients.

Reply:
I added byproduct test for `S_PO4` and `S_NH4`, but to be notice, since we set the water recovery as 1, so the byproduct stream is assigned some system errors extremely close to 0. So the byproduct concentration will be extremely high and does not make much sense.

Reply:
@luohezhiming Maybe just add an `assert value(m.fs.unit.byproduct.flow_vol[0])` so that it's more clear why the byproduct values are so high?